server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/BazelBspServerBuildManager.java

```diff
@@ -64,13 +64,32 @@ public BazelBspServerBuildManager(
     return bazelBspQueryManager.getSourceItems(rule, label);
   }
 
-  public String getSourcesRoot(String uri) {
-    List<String> root =
-        Constants.KNOWN_SOURCE_ROOTS.stream().filter(uri::contains).collect(Collectors.toList());
-    return bazelData.getWorkspaceRoot()
-        + (root.size() == 0
-            ? ""
-            : uri.substring(1, uri.indexOf(root.get(0)) + root.get(0).length()));
+  public String getSourcesRoot(URI sourceUri) {
+    Path sourcePath = Paths.get(sourceUri);
+    FileSystem fs = FileSystems.getDefault();
+    PathMatcher sourceRootPattern = fs.getPathMatcher(
+            "glob:**/"
+                    + "{main,test,tests,src,3rdparty,3rd_party,thirdparty,third_party}/"
+                    + "{resources,scala,java,kotlin,jvm,proto,python,protobuf,py}");
+    PathMatcher defaultTestRootPattern = fs.getPathMatcher("glob:**/{test,tests}");
+    Path sourceRootGuess = null;
+    for (PathMatcher pattern : new PathMatcher[]{sourceRootPattern, defaultTestRootPattern}) {
+      sourceRootGuess = approximateSourceRoot(sourcePath, pattern);
+      if (sourceRootGuess != null) break;
+    }
+    if (sourceRootGuess == null) {
+      return sourcePath.getParent().toString();
+    }
+    return sourceRootGuess.toAbsolutePath().toString();
+  }
+
+  private Path approximateSourceRoot(Path dir, PathMatcher matcher) {
```

Comment:
i believe that it would be better to use `Optional` and return `Empty` instead of `null`. Also, it would be more efficient to change recursion into a loop

Reply:
Removed the recursion. Also, I think it's okay to use null here, because this is just a private method, and only a few lines long. On the other hand, the `getParent` method may return null anyway, and I'd rather not introduce Java optionals just for the sake of it.